File: admin/db/options.php
<?php

class TB_Options {
	public function get($name){
		global $tbdb;
		$sql = 'SELECT value FROM options where name=?';
		if($stmt = $tbdb->prepare($sql)){
			$stmt->bind_param('s', $name);
			$stmt->execute();
			$stmt->bind_result($value);
			$stmt->fetch();
			$stmt->close();
			return $value;
		}

		return false;
	}

	public function set($name, $value){
		global $tbdb;
		if($this->has($name)){
			$sql = 'UPDATE options SET value=? WHERE name=?';
			if($stmt = $tbdb->prepare($sql)){
				$stmt->bind_param('ss', $value, $name);
				return $stmt->execute();
			}
		} else {
			$sql = 'INSERT INTO options (name,value) VALUES (?,?)';
			if($stmt = $tbdb->prepare($sql)){
				$stmt->bind_param('ss', $name, $value);
				return $stmt->execute();
			}
		}

		return false;
	}		

	public function has($name){
		global $tbdb;
		$sql = 'SELECT name FROM options WHERE name=?';
		if($stmt = $tbdb->prepare($sql)){
			$stmt->bind_param('s', $name);
			$stmt->execute();
			$ret = $stmt->get_result();
			return $ret!==false && $ret->num_rows>0;
		}

		return false;
	}

	public function del($name) {
		global $tbdb;
		$sql = 'DELETE FROM options WHERE name=?';
		if($stmt = $tbdb->prepare($sql)){
			$stmt->bind_param('s', $name);
			return $stmt->execute();
		}

		return false;
	}
}

File: server/options.go
package main

import (
	"fmt"
	"log"
	"strconv"

	"github.com/movsb/taoblog/server/modules/taorm"
)

type OptionManager struct {
}

type Option struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

func newOptionsModel() *OptionManager {
	return &OptionManager{}
}

func (o *OptionManager) Has(tx Querier, name string) error {
	query := `SELECT name FROM options WHERE name=? LIMIT 1`
	val := ""
	row := tx.QueryRow(query, name)
	return row.Scan(&val)
}

func (o *OptionManager) Get(tx Querier, name string) (string, error) {
	query := `SELECT value FROM options WHERE name=? LIMIT 1`
	row := tx.QueryRow(query, name)
	val := ""
	err := row.Scan(&val)
	return val, err
}

func (o *OptionManager) GetDef(tx Querier, name string, def string) string {
	val, err := o.Get(tx, name)
	if err == nil {
		return val
	}
	return def
}

func (o *OptionManager) GetDefInt(tx Querier, name string, def int64) int64 {
	val, err := o.Get(tx, name)
	if err == nil {
		num, _ := strconv.ParseInt(val, 10, 64)
		return num
	}
	return def
}

func (o *OptionManager) Set(tx Querier, name string, val interface{}) error {
	strVal := fmt.Sprint(val)

	query := ""
	var err error

	if o.Has(tx, name) == nil {
		query = `UPDATE options SET value=? WHERE name=? LIMIT 1`
		_, err = tx.Exec(query, strVal, name)
	} else {
		query = `INSERT INTO options (name,value) VALUES (?,?)`
		_, err = tx.Exec(query, name, strVal)
	}
	log.Println(query)
	return err
}

func (o *OptionManager) Del(tx Querier, name string) error {
	query := `DELETE FROM options WHERE name=? LIMIT 1`
	_, err := tx.Exec(query, name)
	return err
}

func (o *OptionManager) List(tx Querier) ([]*Option, error) {
	var items []*Option
	query := `SELECT name,value FROM options`
	if err := taorm.QueryRows(&items, tx, query); err != nil {
		return nil, err
	}
	return items, nil
}
